src/System.Private.CoreLib/shared/System/String.Comparison.cs

```diff
@@ -65,112 +66,7 @@ private static unsafe int CompareOrdinalHelper(string strA, string strB)
-#if BIT64
-                while (length >= 12)
-                {
-                    if (*(long*)a != *(long*)b) goto DiffOffset0;
-                    if (*(long*)(a + 4) != *(long*)(b + 4)) goto DiffOffset4;
-                    if (*(long*)(a + 8) != *(long*)(b + 8)) goto DiffOffset8;
-                    length -= 12; a += 12; b += 12;
-                }
-#else // BIT64
-                while (length >= 10)
-                {
-                    if (*(int*)a != *(int*)b) goto DiffOffset0;
-                    if (*(int*)(a + 2) != *(int*)(b + 2)) goto DiffOffset2;
-                    if (*(int*)(a + 4) != *(int*)(b + 4)) goto DiffOffset4;
-                    if (*(int*)(a + 6) != *(int*)(b + 6)) goto DiffOffset6;
-                    if (*(int*)(a + 8) != *(int*)(b + 8)) goto DiffOffset8;
-                    length -= 10; a += 10; b += 10; 
-                }
-#endif // BIT64
-
-                // Fallback loop:
-                // go back to slower code path and do comparison on 4 bytes at a time.
-                // This depends on the fact that the String objects are
-                // always zero terminated and that the terminating zero is not included
-                // in the length. For odd string sizes, the last compare will include
-                // the zero terminator.
-                while (length > 0)
-                {
-                    if (*(int*)a != *(int*)b) goto DiffNextInt;
-                    length -= 2;
-                    a += 2;
-                    b += 2;
-                }
-
-                // At this point, we have compared all the characters in at least one string.
-                // The longer string will be larger.
-                return strA.Length - strB.Length;
-
-#if BIT64
-            DiffOffset8: a += 4; b += 4;
-            DiffOffset4: a += 4; b += 4;
-#else // BIT64
-                // Use jumps instead of falling through, since
-                // otherwise going to DiffOffset8 will involve
-                // 8 add instructions before getting to DiffNextInt
-                DiffOffset8: a += 8; b += 8; goto DiffOffset0;
-                DiffOffset6: a += 6; b += 6; goto DiffOffset0;
-                DiffOffset4: a += 2; b += 2;
-                DiffOffset2: a += 2; b += 2;
-#endif // BIT64
-
-            DiffOffset0:
-                // If we reached here, we already see a difference in the unrolled loop above
-#if BIT64
-                if (*(int*)a == *(int*)b)
-                {
-                    a += 2; b += 2;
-                }
-#endif // BIT64
-
-            DiffNextInt:
-                if (*a != *b) return *a - *b;
-
-                DiffOffset1:
-                Debug.Assert(*(a + 1) != *(b + 1), "This char must be different if we reach here!");
-                return *(a + 1) - *(b + 1);
-            }
+            return SpanHelpers.SequenceCompareTo(ref strA.GetRawStringData(), strA.Length, ref strB.GetRawStringData(), strB.Length);
```

Comment:
Does this have any impact positive or negative on perf, in particular for short strings?

Also, I believe call sites to this were already checking the first character. Is that still needed? Is the assert above still valid?

I love the reduction in code, just want to make sure we understand any unanticipated consequences.

Reply:
Cleaned up the call sites as they were making more calls to the throw check than they needed to; but they are big switch methods, so is doubtful they will inline (which the comment it about), the `_firstChar` check is without bounds check so quite light, so probably remains an easy optimization to skip the method call; if its a common case.

Will get some numbers.

Reply:
Actually one of the callsites might inline; but then the first char check would also inline, but SequenceCompareTo definitely won't; so it still works.

Reply:
On https://github.com/dotnet/coreclr/pull/22187, changing char in last position of string
```diff
                 Method | Length |      Mean |     Error |    StdDev |
 ---------------------- |------- |----------:|----------:|----------:|
- string_CompareOrdinal |      1 |  3.204 ns | 3.5938 ns | 0.1970 ns |
+ string_CompareOrdinal |      1 |  3.159 ns | 0.2917 ns | 0.0160 ns |

- string_CompareOrdinal |      4 |  6.343 ns | 0.2986 ns | 0.0164 ns |
+ string_CompareOrdinal |      4 |  9.657 ns | 0.6693 ns | 0.0367 ns |

- string_CompareOrdinal |      5 |  6.561 ns | 0.4839 ns | 0.0265 ns |
+ string_CompareOrdinal |      5 |  6.955 ns | 1.4324 ns | 0.0785 ns |

- string_CompareOrdinal |      6 |  6.611 ns | 0.4435 ns | 0.0243 ns |
+ string_CompareOrdinal |      6 |  7.945 ns | 0.4340 ns | 0.0238 ns |

- string_CompareOrdinal |      7 |  7.400 ns | 0.4771 ns | 0.0262 ns |
+ string_CompareOrdinal |      7 |  8.733 ns | 0.6226 ns | 0.0341 ns |

- string_CompareOrdinal |      8 |  7.438 ns | 0.4543 ns | 0.0249 ns |
+ string_CompareOrdinal |      8 |  7.501 ns | 0.1890 ns | 0.0104 ns |

- string_CompareOrdinal |     15 |  6.256 ns | 0.2924 ns | 0.0160 ns |
+ string_CompareOrdinal |     15 |  8.128 ns | 0.1217 ns | 0.0067 ns |

- string_CompareOrdinal |     16 |  6.232 ns | 0.2279 ns | 0.0125 ns |
+ string_CompareOrdinal |     16 |  6.999 ns | 0.2556 ns | 0.0140 ns |

- string_CompareOrdinal |     24 |  9.363 ns | 0.3676 ns | 0.0201 ns |
+ string_CompareOrdinal |     24 |  7.577 ns | 0.4276 ns | 0.0234 ns |

- string_CompareOrdinal |     31 |  9.098 ns | 0.2410 ns | 0.0132 ns |
+ string_CompareOrdinal |     31 |  7.594 ns | 0.4550 ns | 0.0249 ns |

- string_CompareOrdinal |     32 |  9.105 ns | 0.5410 ns | 0.0297 ns |
+ string_CompareOrdinal |     32 |  7.578 ns | 0.2211 ns | 0.0121 ns |

- string_CompareOrdinal |     33 | 10.227 ns | 0.1747 ns | 0.0096 ns |
+ string_CompareOrdinal |     33 |  8.418 ns | 0.3742 ns | 0.0205 ns |

- string_CompareOrdinal |     63 |  9.809 ns | 0.4018 ns | 0.0220 ns |
+ string_CompareOrdinal |     63 |  9.311 ns | 0.7240 ns | 0.0397 ns |

- string_CompareOrdinal |     64 |  9.833 ns | 0.0420 ns | 0.0023 ns |
+ string_CompareOrdinal |     64 |  9.340 ns | 2.9041 ns | 0.1592 ns |

- string_CompareOrdinal |    127 | 15.538 ns | 0.7282 ns | 0.0399 ns |
+ string_CompareOrdinal |    127 | 12.648 ns | 1.3036 ns | 0.0715 ns |

- string_CompareOrdinal |    128 | 15.628 ns | 1.0243 ns | 0.0561 ns |
+ string_CompareOrdinal |    128 | 12.595 ns | 0.3935 ns | 0.0216 ns |

- string_CompareOrdinal |    255 | 23.341 ns | 0.9011 ns | 0.0494 ns |
+ string_CompareOrdinal |    255 | 19.403 ns | 0.2229 ns | 0.0122 ns |

- string_CompareOrdinal |    256 | 23.223 ns | 0.3965 ns | 0.0217 ns |
+ string_CompareOrdinal |    256 | 19.455 ns | 0.5873 ns | 0.0322 ns |

- string_CompareOrdinal |   1023 | 85.174 ns | 4.8935 ns | 0.2682 ns |
+ string_CompareOrdinal |   1023 | 73.952 ns | 0.7607 ns | 0.0417 ns |

- string_CompareOrdinal |   1024 | 85.041 ns | 7.0647 ns | 0.3872 ns |
+ string_CompareOrdinal |   1024 | 73.840 ns | 5.9904 ns | 0.3284 ns |
```

Reply:
The new numbers generally look worse for <= 16.  Is that expected?

Reply:
Yes currently as that's when vectors kick in; though I think I can turn it around.

Will wait till decision made on https://github.com/dotnet/coreclr/pull/22187 though as its complicated to try a coordinate across the two PRs and that's an improvement on current and I don't want to keep adding changes on changes as its already quite big.

Reply:
i.e. `SequenceCompareTo` does char by char; this does long/int unrolls knowing alignment of string; however should be able to xor then find first changed char